### PR TITLE
Improved cold state checkpoints: migrate database index and usage

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR adds a database migration for the archived index root bucket from the index as `slot/SlotsPerArchivedPoint` to `slot`.

See design doc: https://hackmd.prylabs.network/s/SkbAyN4CU#Updating-Database-Indexes

**Which issues(s) does this PR fix?**

Part of solution for #6325. 

**Other notes for review**

This PR ended up with a lot of refactoring to simplify the database API.

- bytesutil.Uint64ToBytes to BigEndian and LittleEndian
- Added a migration for block slot indices to use BigEndian bytes instead of a string representation of slot. This should make for a more accurate, faster, and less storage index bucket.
- Removed / deprecated old buckets and indices via migrations
- Removed SaveLastArchivedIndex, this index is updated on SaveState
- Removed SaveArchivedPointRoot, this index is updated on SaveState
- Renamed references to archived index to archived slot
- Refactored HighestSlotBlocks to HighestSlotBlock, the only usage of this expected a single block to be returned or it would fail. 
- Removed usage of bitfields for purposes of indexing states, now using the state index bucket and iterating through those fields quickly

There are probably more items I forgot to list. 

**Please question everything and review carefully.**

**Testing notes**

I tested this by syncing altona on master, then making a backup of that database. 
Then I would run:
- cp /tmp/backup.db ~/.eth2/beaconchaindata/beaconchain.db
- bazel run //beacon-chain -- --dev --disable-initial-sync-verify-all-signatures --altona --verbosity=debug

I would also run this bash script and check the epoch returned is aligned with the requests and the requests did not take over 2s. Every 64 epochs would be a new checkpoint, so watching this run for a few hundred epochs gave me confidence that the state regen from cold checkpoint was still working.

```bash
for i in {1..2000}  
do
   time curl -s -X GET "http://localhost:3500/eth/v1alpha1/validators/assignments?epoch=$i" -H "accept: application/json" | jq . | grep epoch
done
```

### CODE REVIEW

File: beacon-chain/state/stategen/migrate.go [hunk 1]
```diff
@@ -15,6 +15,21 @@ import (
 // MigrateToCold advances the finalized info in between the cold and hot state sections.
 // It moves the recent finalized states from the hot section to the cold section and
 // only preserve the ones that's on archived point.
+//
+// This method does:
```
Comment (terencechain) [inline beacon-chain/state/stategen/migrate.go:19]: This looks outdated

Comment (prestonvanloon) [reply beacon-chain/state/stategen/migrate.go:19]: Looks like a bad merge, I'll delete it

Comment (prestonvanloon) [reply beacon-chain/state/stategen/migrate.go:19]: oh these were my notes

File: beacon-chain/db/kv/migration_archived_index.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+
```
Comment (rauljordan) [inline beacon-chain/db/kv/migration_archived_index.go:8]:
```suggestion
```

File: beacon-chain/db/kv/migration_archived_index_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"go.etcd.io/bbolt"
+
```
Comment (rauljordan) [inline beacon-chain/db/kv/migration_archived_index_test.go:11]:
```suggestion
```

File: beacon-chain/db/kv/state.go [hunk 4]
```diff
@@ -6,12 +6,12 @@ import (
 
 	"github.com/pkg/errors"
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
-	"github.com/prysmaticlabs/prysm/beacon-chain/db/filters"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+
```
Comment (rauljordan) [inline beacon-chain/db/kv/state.go:11]:
```suggestion
```

**Comment by terencechain** [comment]
I've done some live testing with this. 👍 from me.

What i tested:
2048 slots per archive point:
 - sync to head
 - verified correct amount of states are saved via metrics
 - rpc calls that involves cold state

32 slots per archive point:
 - sync to head
 - verified correct amount of states are saved via metrics
 - rpc calls that involves cold state (much faster than 2048

One minor concern I have with this is there's no clean way of going back. Once this hits `master` or the next `release` there's no going back to the old scheme. Tried to go back and got:
```
[2020-07-15 14:20:21]  INFO initial-sync: Block is not processed error=parent root 0xe681f01260438c9b63329b60b29dc44244181ab639d065f691b2f15622345208 does not match the latest block header signing root in state 0x1d42753193a3eafb78ebf9389f15401b661ce0be824ac45dc34d7ad3e794363d
```
I think that's the risk we have to take. I'm ok with that

**Comment by prestonvanloon** [comment]
@terencechain yeah, there isn't a great way to go about database migrations. It will be a breaking change from alpha.15 to alpha.16. Thankfully, the migration files are rather simple and reversible. If we needed to revert the migration, we could do so with a small tool. 

I think this PR is ready to go!

**Comment by prestonvanloon** [comment]
Good point. I'll update the index to track multiple roots

**Comment by terencechain** [comment]
> This looks good to me. A few minor points which I'm not sure if we want to address now or later
> 
> * `BytesToUint64LittleEndian` is not used
> * `Uint64ToBytesBigEndian` has no direct unit test
> * `BytesToUint64BigEndian` has no direct unit test

**Comment by prestonvanloon** [comment]
This PR is ready to go, but please do not merge until Saturday, after the release cut off.

**Pull request merged**